add tutorial to connect to flux between clusters

vsoch opened the pull request:
This uses a proxy jump in the ssh config.

I don't have this working yet, but wanted to open the PR to clearly lay out what I'm doing (so we can see what I'm doing wrong).

Update: all is working! I was missing `export FLUX_SSH=ssh` and then `--force`

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
Preview: https://flux-framework--192.org.readthedocs.build/en/192/tutorials/lab/ssh.html

chu11 commented:
High level comment, this feels "less than" a tutorial but "more than" a FAQ entry.  Would some type of "how to" section be wise?  Where we could put general help kinda things?

vsoch commented:
@chu11 I disagree with your labeling this as "less than" a tutorial - it walks the reader through a complete process, and I would argue is exactly the right size to grab and maintain attention. If it was a one off command? Then maybe it wouldn't be a tutorial. But it's an entire process with multiple processes and steps, thus it falls nicely here.

We also should not have everything bunched under "FAQ" - it's already too busy. This in particular is specific to LC systems (we cannot guarantee it would work on others we have not tested) and I think is placed exactly where I'd want to find it.

springme commented:
To Al’s suggestion that it’s more than what belongs in a FAQ, maybe there could be levels of Tutorial where some are “How To” tutorials and others are lengthy multi-topic tutorials that may  require a lot more time to view.

If there are lots of them there might be some good ways to group them.


From: Vanessasaurus ***@***.***>
Reply-To: flux-framework/flux-docs ***@***.***>
Date: Monday, February 6, 2023 at 12:30 PM
To: flux-framework/flux-docs ***@***.***>
Cc: Subscribed ***@***.***>
Subject: Re: [flux-framework/flux-docs] add tutorial to connect to flux between clusters (PR #192)


@chu11<https://urldefense.us/v3/__https:/github.com/chu11__;!!G2kpM7uM-TzIFchu!kTeRdDOMiDc_Wcygd7np8N--k84FvPL97CSo3PXJt5p8CELVeGhzDQnW-5AWlDVlXNtY$> I disagree with your labeling this as "less than" a tutorial - it walks the reader through a complete process, and I would argue is exactly the right size to grab and maintain attention. If it was a one off command? Then maybe it wouldn't be a tutorial. But it's an entire process with multiple processes and steps, thus it falls nicely here.

We also should not have everything bunched under "FAQ" - it's already too busy. This in particular is specific to LC systems (we cannot guarantee it would work on others we have not tested) and I think is placed exactly where I'd want to find it.

—
Reply to this email directly, view it on GitHub<https://urldefense.us/v3/__https:/github.com/flux-framework/flux-docs/pull/192*issuecomment-1419709376__;Iw!!G2kpM7uM-TzIFchu!kTeRdDOMiDc_Wcygd7np8N--k84FvPL97CSo3PXJt5p8CELVeGhzDQnW-5AWlJ56Z1_W$>, or unsubscribe<https://urldefense.us/v3/__https:/github.com/notifications/unsubscribe-auth/ABDVJUGE2NXTFC5ON5QJIIDWWFNPLANCNFSM6AAAAAAUTCGKAQ__;!!G2kpM7uM-TzIFchu!kTeRdDOMiDc_Wcygd7np8N--k84FvPL97CSo3PXJt5p8CELVeGhzDQnW-5AWlDcVcZw9$>.
You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

chu11 commented:
> I do agree there should be something between FAQ and tutorial, and for general things, although I'm not sure what that looks like. 

Perhaps just a bullet under the tutorials section for "General How Tos" or something could suffice for now.  And if it grows large enough we split out into another page.

I now realize that you added this under the "Lab tutorials" section.  In my mind it was more "general" at first, but perhaps it swerves lab specific a bit.

chu11 commented on tutorials/lab/ssh.rst:
```diff
@@ -0,0 +1,99 @@
+----------------------
+
+
+First, let's create the allocation on the first cluster.
```
Would it be possible to generalize it a bit?  On clusters with native flux you might typically do  ... on clusters with slurm ... on lassen see <link>?

vsoch replied:
@grondo this command?

```
$ jsrun -a 1 -c ALL_CPUS -g ALL_GPUS -n ${NUM_NODES} --bind=none --smpiargs="-disable_gpu_hooks" flux start
```
This definitely calls for a "Rosetta stone" of starting flux commands - e.g., I could imagine the same but for slurm, or another job manager. I think there should be one place where someone can go and see them all side by side (although separate from this PR). What I can do is put this link / reference in the tutorial, and I'm still thinking about how we can structure a more general set of mini tutorials.

vsoch replied:
@chu11 I agree we want this "how to start flux" across different places somewhere, maybe not here but definitely somewhere, along with other useful commands and different contexts / quick reference for running them.

chu11 replied:
i thought we already had the how to launch a sub-instance on FAQ, but I guess not ... seems like we should definitely add something.

vsoch replied:
@grondo I'm testing jsrun, and maybe I'm doing it wrong, but this doesn't work:

```
$ export NUM_NODES=1
$ jsrun -a 1 -c ALL_CPUS -g ALL_GPUS -n ${NUM_NODES} --bind=none --smpiargs="-disable_gpu_hooks" flux start
Error: It is only possible to use js commands within a job allocation unless CSM is running
02-06-2023 13:43:54:480 82493 main: Error initializing RM connection. Exiting.
```
So I followed the message and got an allocation first:

```bash
$ jsrun -a 1 -c ALL_CPUS -g ALL_GPUS -n ${NUM_NODES} --bind=none --smpiargs="-disable_gpu_hooks" flux start
flux-broker: stdin is not a tty - can't run interactive shell
```
(and no flux instance is started) is there another command on that page I should be using? https://flux-framework.readthedocs.io/en/latest/tutorials/lab/coral.html#launching-spectrum-mpi-within-flux

vsoch replied:
@chu11 I'm trying to get that working right now (see above) no luck so far! I agree we need an entire general page on this too.

vsoch replied:
Ping @grondo I need help with the jsrun command (doesn't seem to work) when you have a moment! I think now #194 is waiting on this to merge first.

chu11 replied:
for the purposes of this example, I don't think it needs to be jsrun, right?  Could it just be replaced with a slurm example?

vsoch replied:
It definitely could, I just don't know what that looks like, haha. I wanted to test it too to ensure it works, but if someone else knows it works for sure I'd be happy to add that!

I agree since we generalized it slurm would be better, I don't know what a jsrun is :)

chu11 replied:
This is from the lab's tutorial that @ryanday36 wrote.

```
>salloc -N4 --exclusive
salloc: Granted job allocation 321075
salloc: Waiting for resource configuration
salloc: Nodes opal[63-66] are ready for job
>srun -N4 -n4 --pty --mpibind=off flux start
>flux mini run -N4 hostname
opal63
opal64
opal65
opal66
```

chu11 replied:
Another way from a manpage is: `srun --pty -N8 flux start`

I'm actually struggling to find a way to do it A) without a pty and B) allow the user to easily get the FLUX_URI ...

vsoch replied:
Ah I see - so it's getting an allocation first still, but then running flux start with srun and then you can interact. Let me update the tutorial to use that, and if I can get access to a slurm cluster sometime with flux I can test again.

chu11 replied:
> I'm actually struggling to find a way to do it A) without a pty and B) allow the user to easily get the FLUX_URI ...

Ahh `flux uri` has a slurm resolver based on a searching pids.

grondo replied:
I don't know `jsrun` usage either. You can also connect directly to a flux instance running under Slurm with `flux proxy slurm:JOBID`. This is documented in [Working with Flux job hierarchies](https://flux-framework.readthedocs.io/en/latest/jobs/hierarchies.html)

vsoch commented:
okay so here is an intermediate idea - what if we have "Command Tutorials" e.g., a section here:

![image](https://user-images.githubusercontent.com/814322/217085962-8c47ff62-dea5-477d-a290-bb485f7b8e5f.png)

And then under there, we have each of the commands, and flux proxy would be one of them (and I'm happy to try and generalize the current doc here and move away from being under lab tutorials, although I cannot guarantee it would be the same on other lab clusters). In these Command tutorials we would generally try to show a lot of examples and contexts for doing something, and if there is a "source of truth" in one of the rc guides or similar, we'd link to it. 

I think under Tutorials is the right place - I'm looking at FAQ and it's really busy - at best someone is going to find matching content here via a search. The other place I go is "Quickstart" but given that is supposed to be quick, I'm hesitant to add more content there (although I do think we can work on it more). So that leaves us with Guides and Tutorials, and a third option to add something that doesn't exist yet. Guides (imho) are complete guides for an entire user base, so I don't think it goes there. It could fall under a new category "Command Guide?" or similar, but I think that could also fit under tutorials since we just have two types at the moment. What do others think?

chu11 commented:
> # This is the main Command Tutorials page
> - `flux proxy`: "I want to connect to a flux instance across clusters with ssh
> - `flux start`: "I want to start my own flux allocation/instance to launch jobs"
> - `flux mini submit/run`: "I want to run or submit a job to a flux instance"

I like this idea, we could have different "levels" of tutorials too.  I'm imaging

> - `flux proxy`: "send commands to a flux instance you've started"
> - "flux proxy": "send commands to a flux instance across clusters using ssh"

vsoch commented:
> I like this idea, we could have different "levels" of tutorials too. I'm imagining

Yes!! OK let me start us on this path - the pages will be a bit empty to start (one tutorial!) but we have to start somewhere!

vsoch commented:
okay new "Commands Tutorials" are added! https://flux-framework--192.org.readthedocs.build/en/192/tutorials/commands/index.html But we should not merge yet because the jsrun command didn't work for me (see comment above) https://github.com/flux-framework/flux-docs/pull/192#discussion_r1097954413

chu11 commented on tutorials/commands/ssh-across-clusters.rst:
```diff
@@ -0,0 +1,167 @@
+
+    $ local:///var/tmp/flux-MLmxy2/local-0    
+
+The 
+
```
errant `The` :-)

vsoch replied:
How did he escape!??! :runner:

chu11 commented:
high level comment as I'm reading the tutorial, maybe put the hostname into the prompts, like:

`lassen:~$`

or something.  So there's a little bit that differentiates the output from when you're typing on lassen vs quartz (as some output like `flux resource list` is identical).

vsoch commented:
suggested todos:

 - for this PR: get the working command for flux start on lassen (so I can test)
 - I can start an issue that lists the guides / "command tutorials" we want to write (and post ideas here)

Going for a run backinabit.

vsoch commented:
Updated preview: https://flux-framework--192.org.readthedocs.build/en/192/tutorials/commands/ssh-across-clusters.html

 - I added the srun example to start the instance
 - I generalized the name of the first cluster so it's not heavy with our cluster names!

chu11 commented on tutorials/commands/ssh-across-clusters.rst:
```diff
@@ -0,0 +1,156 @@
+    }
+
+And that's it! With this strategy, it should be easy to interact with Flux instances from
+two resources where ssh is supported. If you have any questions, please `let us know <https://github.com/flux-framework/flux-docs/issues>`_.
```
> If you have any questions, please `let us know <https://github.com/flux-framework/flux-docs/issues>`_.

Instead of having something like this in every tutorial, perhaps we just need a header/footer comment or something within the "tutorials" page?

chu11 replied:
or ... if people get to this page via search engines, perhaps it good to have at the bottom.

vsoch replied:
But then if they are in a specific tutorial, they wouldn't see it right? It's important (I think) to be at the bottom of the page so the reader might glance through the tutorial, and then feel like they are missing something / have a question and immediately see that.

chu11 commented on tutorials/commands/ssh-across-clusters.rst:
```diff
@@ -0,0 +1,156 @@
+SSH across clusters
+===================
+
+Let's say you want to create a Flux instance in an allocation on a cluster (e.g., let's say out first cluster is "noodle") 🍜️
+and then connect to it via ssh from another cluster (let's say our second cluster is called quartz). This is possible with the right
```
should quartz be in quotes too? like noodle?

vsoch replied:
sure, that would be consistent!

vsoch commented on tutorials/commands/ssh-across-clusters.rst:
```diff
@@ -0,0 +1,156 @@
+.. code-block:: sh
+
+    # slurm specific
+    $ salloc -N4 --exclusive
+    $ srun -N4 -n4 --pty --mpibind=off flux start
```
I thought we were on some general noodle node (that is part of the allocation?)

vsoch replied:
okay I added a comment at the top that they are on a login node, and then when they hit "noodle" they have their allocation.

chu11 replied:
My reasoning for the prompt change is b/c of your ssh config below.  `noodle` appears to be the login node, while `noodle220` is the node that you were allocated to run your job.  So it may not be clear which node you're actually on with just `noodle:~$`??

For example, w/ my personal prompt:

```
opal186|/g/g0/achu 44>salloc -N4 -ppbatch
salloc: Granted job allocation 321078
salloc: Waiting for resource configuration
salloc: Nodes opal[63-66] are ready for job

opal63|/g/g0/achu 21>srun hostname
opal63
opal66
opal65
opal64
```

you'll notice I was on opal186, the login node, and then `salloc` dropped me into a shell on `opal63`.  So if I were to setup the ssh config, I would think I should set it up for `opal186` and `opal63`.

vsoch replied:
I changed it so the login node is just empty (no name) and I explicitly state we are on the login node. Then I state we are on the allocation and just use `noodle:~$` to say we are on the allocation (and the specific node largely doesn't matter). I thought it looked nicer without the number so I left it out.

vsoch commented on tutorials/commands/ssh-across-clusters.rst:
```diff
@@ -0,0 +1,156 @@
+
+.. code-block:: sh
+
+    $ flux job list  | jq
+    {
+      "id": 2272725565440,
+      "userid": 34633,
+      "urgency": 16,
+      "priority": 16,
+      "t_submit": 1675713045.009863,
+      "state": 16,
+      "name": "sleep",
+      "ntasks": 1,
+      "nnodes": 1,
+      "ranks": "2",
+      "nodelist": "noodle220",
+      "expiration": 1676317845,
+      "t_depend": 1675713045.009863,
+      "t_run": 1675713045.0290241,
+      "annotations": {
+        "sched": {
+          "queue": "default"
+        }
+      }
+    }
```
Does flux jobs give the same json output?

chu11 replied:
no, it gives the pretty output, example from the other PR

```
    $ flux jobs
       JOBID  USER     NAME       ST NTASKS NNODES     TIME INFO
    ƒSUEFPDH  fluxuser my_other_s  R      1      1   1.842s 
    ƒM5k8m7m  fluxuser my_compute  R      4      2   3.255s
```

`flux job list` is really only for testing, and was actually disabled for general use a few releases ago.

```
>src/cmd/flux start
>flux job list
This is not the command you are looking for. Try flux-jobs(1).
```

vsoch replied:
Gotcha! Let me see if I can update it.

vsoch replied:
_This is not the droid that you seek..._

chu11 commented on tutorials/commands/ssh-across-clusters.rst:
```diff
@@ -0,0 +1,138 @@
+    noodle222
+    noodle223
+
+And you might adopt this command to be more specific to your resource manager. E.g., slurm uses srun.
```
i would probably just remove this line, not sure if it serves much purpose at this point?  The `flux mini run hostname` is to illustrate that we're inside a flux instance.

vsoch replied:
Remove the line about adopting the command? Or the flux mini run? I think both are important to have:

- the first recognizes that "if your hpc manager is different, you'll need to adopt this command"
- the second gives the reader a way to sanity check they are where they think they are, and a quick preview for how to run a very simple command (tying two different tutorials - the submit and ssh together).

We can see what others think. I don't think it hurts to have it, and actually might help.

chu11 replied:
ohhh, you're referencing the `srun -N4 -n4 --pty --mpibind=off flux start` line from earlier ... i thought it was referencing the `flux mini run hostanme` example, which was why it seemed weird why you were saying it.

Perhaps we just need to move it up above the `flux mini run hostname` line.  Like

```
And then we get our allocation!  This example launches flux with slurm (i.e. salloc / srun), you may need to adjust for a different resrouce manager
```

Also, I think you meant "adapt" not "adopt"?

vsoch replied:
yes we can move it earlier!

And good catch... I'll fix that in a minute (having some cocoa). Side note - is there a signup for adopting nodes? I can adopt them to live with my pet pusheens.

chu11 commented on tutorials/commands/ssh-across-clusters.rst:
```diff
@@ -0,0 +1,138 @@
+And you might adopt this command to be more specific to your resource manager. E.g., slurm uses srun.
+After you run flux start, you are inside of a Flux instance! We generally want to launch 
+this instance as a parallel job under the native resource manager, rather than get an allocation and just run flux start.
+The reason is because the test instance ignores native resource allocation and starts N brokers in place.
```
> The reason is because the test instance ignores native resource allocation and starts N brokers in place.

I think Jim was referring to when you were doing `flux start -s4` before.  In this case its not a test instance.  I think you can just remove this sentence.

vsoch replied:
Is that not an interesting bit of knowledge to include? I thought it was.

chu11 replied:
In an earlier iteration of your doc you used `flux start --test-size=4` (or maybe -s4)   And that's what Jims comment was about earlier when he said:

> Generally, you'd want to launch the instance as a parallel job under the native resource manager, rather than get an allocation and use flux-start --test-size=N. The test instance ignores the native resource allocation and just starts N brokers in place.

Now with `srun -N4 -n4 --pty --mpibind=off flux start`, this is a real flux instance against the native resource manager, not a test instance.   So the text seems out of place.

vsoch replied:
okay, I'm good to remove then.

chu11 reviewed (approved):
LGTM!  Thanks for getting this new batch of tutorials / docs starts.

vsoch commented:
@grondo I think we can merge with your blessing? And then unblock @cmoussa1.

vsoch commented:
Note sure if I'm allowed to do this, but since @chu11 approved, I set that label.

chu11 commented:
yes, once approved you can set MWP.  Sometimes if the reviewer wants another to skim before you set MWP we'll say "oh hey such and such should take a look though b/c I'm not an expert on FOO".

vsoch commented:
Thank you for approving!

vsoch commented:
@chu11 I'm not great with using flux commands, but if you want to point me at a particular example from the workflows repo (the one we are going to archive), I can work on integrating that here next. It will be good for the docs and my learning. :)

chu11 commented:
In my opinion we may eventually want to add python equivalents to the CLI tutorials, like "oh hey, you wanna do this in python, do it this way" .... like there could be a python job-submit one after PR #194 .  Then we can eventually combine them all into an advanced one, like the python equivalent to my PR #195.

Problem is the python scripts in flux-workflow-examples probably aren't up to date.  It would not surprise me if some of them didn't work at all.

vsoch commented:
I'm actually more familiar with the python cli (given the small amount of work I've done with it) so that might be a good place for me to start. I'll see if I can add sphinx-gallery too so the tutorials render with the current version of flux on build. I'll need to add a .devcontainer environment for us to work from.

Will start soon, latest next week!

vsoch commented:
Okay, step 1: https://github.com/flux-framework/flux-docs/pull/201
